tests/trainer/test_config_validator.py

```diff
@@ -156,3 +158,38 @@ def test_trainer_manual_optimization_config(tmpdir):
     trainer = Trainer(accumulate_grad_batches=2)
     with pytest.raises(MisconfigurationException, match="Automatic gradient accumulation is not supported"):
         trainer.fit(model)
+
+
+def test_invalid_setup_method():
+    """Test error message when `setup` method of `LightningModule` or `LightningDataModule` is not defined
+    correctly."""
+
+    class CustomModel(BoringModel):
+        def setup(self):
+            pass
+
+    class CustomDataModule(BoringDataModule):
+        def setup(self):
+            pass
+
+    class CustomBoringCallback(Callback):
+        def setup(self, pl_module, trainer):
+            pass
+
+    fit_kwargs = [
+        {"model": CustomModel(), "datamodule": BoringDataModule()},
+        {"model": BoringModel(), "datamodule": CustomDataModule()},
+    ]
+
+    for kwargs in fit_kwargs:
+        trainer = Trainer(fast_dev_run=True)
+
+        with pytest.raises(MisconfigurationException, match="does not have a `stage` argument"):
+            trainer.fit(**kwargs)
+
+    trainer = Trainer(fast_dev_run=True, callbacks=[CustomBoringCallback()])
+    model = BoringModel()
+    dm = BoringDataModule()
+
+    with pytest.raises(MisconfigurationException, match="does not have a `stage` argument"):
+        trainer.fit(model, datamodule=dm)
```

Comment:
The dm is not necessary here

```suggestion

    with pytest.raises(MisconfigurationException, match="does not have a `stage` argument"):
        trainer.fit(model)
```